src/utils/generator.js

```diff
@@ -240,6 +240,25 @@ export class Generator {
     return this.realm.intrinsics.undefined;
   }
 
+  /* Emits a declaration for an identifier into the generator at the current point in time
+     and initializes it with the given abstract value, thus ensuring that the corresponding
+     expression is evaluated at a particular point in time. Returns an abstract value that
+     refers to the value via the newly declared identifier. */
+  emitAbstract(
```

Comment:
Hm, this indicates that there are abstract values (the `val` coming in here) which are not pure. That seems wrong. In the past, the `isPure` designation was rather associated with the buildNode itself.

Reply:
Actually, the values are pure. The point is that they must be computed at a particular point in time. I'm moving this to AbstractValue and I'm renaming it to makeTemporal.

In general, the notion that an abstract value need not be pure is deeply troubling. We should revisit the whole issue, but not in this pull request.